---
title: 'Protocol for the Processing and Downstream Analysis of Phosphoproteomics Data with PhosR'
author:
- name: Hani Jieun Kim, Taiyun Kim, Di Xiao and Pengyi Yang
  affiliation:
  - School of Mathematics and Statistics, The University of Sydney
  - Computational Systems Biology Group, Children’s Medical Research Institute, Faculty of Medicine and Health, The University of Sydney
  - Charles Perkins Centre, The University of Sydney

output:
  BiocStyle::html_document:
    toc_float: true
package: BiocStyle
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Step-by-Step Method Details

## Step 1: Installing PhosR
Full installation of PhosR includes downloading the PhosR package from GitHub or BioConductor. To use the latest developmental version of PhosR, install from GitHub. An example of how to perform all steps of this protocol using real data is available on the project GitHub at https://pyanglab.github.io/PhosR/index.html.

Install PhosR by running the following code:
```{r message = FALSE}
if(!require(devtools)){ 
 install.packages("devtools") # If not already installed 
}

#devtools::install_github("PYangLab/PhosR", 
#                         build_opts = c("--no-resave-data", "--no-manual"), 
#                         build_vignettes = TRUE)
library(PhosR)
```

Whilst this STAR protocol provides an in-depth step-by-step tutorial for running PhosR with tips and suggestions, you can browse through the streamlined version of the vignette by running the following code. 
```{r}
#utils::browseVignettes("PhosR")
```

We will also need to load the SummarizedExperiment package. 
```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
```

## Step 2: Creating the PhosphoExperiment Object
To increase the usability of PhosR functions, we implement a “PhosphoExperiment” (ppe) object based on the “SummarizedExperiment” class. To create the PhosphoExperiment object, you will be required to provide a quantification matrix where columns refer to samples and rows refer to phosphosites. Additional annotation labels for phosphosites should be provided alongside the matrix, including the phosphosite residue and position, “sequence window” that captures the amino acids flanking the phosphorylation sites, and the official gene symbol of the host protein in capital letters. Here, we will show the basic steps for generating a PhosphoExperiment object for an exemplar phosphoproteomic data from two liver cell lines, FL83B and Hep 1-6 cells (Humphrey et al., 2015). Briefly, this dataset contains the phosphoproteomic quantifications of mouse hepatocyte cell lines that were treated with either PBS (mock) or insulin. Each condition includes six biological replicates. 

Load the phosphoproteomic quantification matrix pre-processed.
```{r}
# Load dataset
load("../Data/PXD001792_raw_hepatocyte.RData", verbose = TRUE)

# Extract the columns of interests before start 
samples = colnames(PXD001792_raw_hepatocyte)
PXD001792_raw_hepatocyte <- PXD001792_raw_hepatocyte[,c(grep("_Control_\\d$", samples),
                                                        grep("_Ins_\\d$", samples))]
PXD001792_raw_hepatocyte = PXD001792_raw_hepatocyte[, sort(colnames(PXD001792_raw_hepatocyte))]
```

Create PhosphoExperiment Object
```{r}
ppe <- PhosR::PhosphoExperiment(assays = list(Quantification = as.matrix(PXD001792_raw_hepatocyte)))
```

Add site annotations to PhosphoExperiment object
```{r}
splitSites = strsplit(rownames(ppe), "~")
GeneSymbol <- toupper(as.character(sapply(splitSites, function(x) x[[2]])))
Residue <- as.character(gsub("[0-9]", "", sapply(splitSites, function(x) x[[3]])))
Site <- as.numeric(gsub("[STY]","",sapply(splitSites, function(x) x[[3]])))
Sequence <- as.character(sapply(splitSites, function(x) x[[4]]))

ppe@GeneSymbol <- GeneSymbol
ppe@Residue <- Residue
ppe@Site <- Site
ppe@Sequence <- Sequence
```

Alternatively, we can create PhosphoExperiment object as following
```{r}
ppe <- PhosR::PhosphoExperiment(assays = list(Quantification = as.matrix(PXD001792_raw_hepatocyte)), Site = Site, GeneSymbol = GeneSymbol, Residue = Residue, Sequence = Sequence)
```

Lastly add colData information
```{r}
sample_name = strsplit(gsub("^Intensity.", "", colnames(ppe)), "_")

df = S4Vectors::DataFrame(
    cellline = sapply(sample_name, "[[", 1),
    condition = sapply(sample_name, "[[", 2),
    replicate = sapply(sample_name, "[[", 3)
)
rownames(df) = colnames(ppe)
colData(ppe) = df
```

Have a quick glance of the object
```{R}
ppe
dim(ppe)
```

## Step 3: Data pre-processing
The presence of missing values in quantitative phosphoproteomics reduces the completeness of data. Whilst imputation has been widely applied to handle missing values, it remains a major challenge when analyzing phosphoproteomic data and has significant impact on downstream analysis such as normalization. PhosR provides users with greater flexibility for imputation with imputation functions such as ‘scImpute’ and ‘tImpute’. Here, we will go through each function step by step to demonstrate their use in imputing phosphoproteomic data. 

Log transformation
```{r}
logmat = log2(assay(ppe, "Quantification"))
logmat[is.infinite(logmat)] = NA
assay(ppe, "Quantification") = logmat
```

Filtering using function `selectGrps`
```{r}
grps <- paste0(colData(ppe)$cellline, "_", colData(ppe)$condition)
ppe <- selectGrps(ppe, grps, 0.5, n=1)
dim(ppe)
```
Note: The ‘Quantification’ assay in PhosphoExperiment object is now filtered of phosphosites with high number of missing values. Please ensure you have saved the original matrix under a different name if you would like to keep the unfiltered matrix for future analysis.

### Imputation
PhosR enables site- and condition-specific imputation. Here, for each phosphosite in each condition, we impute its missing values in that condition (if any) using site- and condition-specific imputation if the quantification rate within that condition is equal to or greater than a desired percentage (such as 50% below). 
```{r}
set.seed(123)
ppe <- scImpute(ppe, 0.5, grps)
ppe
```

Lastly, we can impute the remaining sites using tail-based imputation 
```{r}
ppe <- tImpute(ppe, assay = "imputed")
```
At this stage, the imputed quantification matrix is stored as an assay called ‘imputed’ in the PhosphoExperiment object.

```{r}
ppe <- medianScaling(ppe, scale = FALSE, assay = "imputed")
ppe
```
The scaled quantification matrix can be found as the ‘scaled’ matrix in the PhosphoExperiment object.

After imputation, we take the ratios for each of the two cell lines
```{r}
FL83B.ratio <- assay(ppe, "scaled")[, grep("FL83B_", colnames(ppe))] - 
rowMeans(assay(ppe, "scaled")[,grep("FL83B_Control", colnames(ppe))])
Hepa.ratio <- assay(ppe, "scaled")[, grep("Hepa1.6_", colnames(ppe))] - 
rowMeans(assay(ppe, "scaled")[,grep("Hepa1.6_Control", 
colnames(ppe))])
assay(ppe, "scaled") <- cbind(FL83B.ratio, Hepa.ratio)
```

We will now save this fully processed matrix for use later
```{r}
PXD001792_ppe_hepatocyte = ppe 
#save(PXD001792_ppe_hepatocyte, file = "../Data/PXD001792_ppe_hepatocyte.RData")
```

CRITICAL: After imputation, data from label-free quantification are typically converted to ratios before subsequent analysis. In contrast to label-free data, you do not need to take ratios of phosphoproteomic data derived from SILAC quantification since the values are inherently ratios (typically with respect to the control sample). The tail-based imputation (Beck et al., 2015) is designed specifically for label-free data (such as the one used in our example) and is not applicable to SILAC data.

## Step 4: identifying stably phosphorylated sites
Several commonly used data normalisation approaches such as the 'removal of unwanted variation” (RUV) (Gagnon-Bartsch and Speed, 2012) require a set of internal standards whose expression are known to be unchanged in the samples measured. This is a challenge for phosphoproteomic data since phosphorylation is a highly dynamic biochemical process. Identifying a set of stably phosphorylated sites (SPSs) is a unique feature of PhosR which enables users to identify context-dependent sets of SPSs. We also included a set of 100 SPSs as a resource, identified from multiple high-quality datasets generated from different cell types and experimental conditions (Kim et al., 2021). As an example, we will use three datasets to demonstrate how SPSs can be identified from multiple phosphoproteomic datasets. Users may wish to replace the example datasets with their own collection of datasets.

Load datasets
```{r}
load("../Data/PXD010621_ppe_ESC.RData", verbose = TRUE) 
load("../Data/PXD003631_ppe_adipocyte.RData", verbose = TRUE) 
load("../Data/phospho_ppe_adipocyte.RData", verbose = TRUE) 

# Simplify names of datasets
ppe1 <- PXD010621_ppe_ESC
ppe2 <- PXD003631_ppe_adipocyte
ppe3 <- phospho_ppe_adipocyte
```

Prepare inputs to the function
```{r}
# Make a list of all PhosphoExperiment objects
ppe.list <- list(ppe1, ppe2, ppe3)

# Make a vector of the selected assays in each of the PhosphoExperiment objects
assays <- "Quantification"

# Make a list of grouping information of each dataset
cond.list <- list(
    grp1 = gsub("_.+", "", colnames(ppe1)),
    grp2 = gsub("_r[0-9]", "", colnames(ppe2)),
    grp3 = colnames(ppe3))
```
Note: If inputs are not PhosphoExperiment objects, please transfer the data format to a PhosphoExperiment object referring to the instructions in Step 2. Example datasets are processed datasets after filtering, normalization and ratio converting. Please refer to Step 4 for data normalization guidance.

Identifying SPSs by calling getSPS()
```{r}
inhouse_SPSs <- getSPS(ppe.list, assays = assays, conds = cond.list)
```

Note: Please note that presence of sufficient overlapped phosphosites identified from the input datasets is critical to identify SPSs. You will receive an error if the number of overlapped sites is fewer than 200 in at least two datasets or fewer than 1000 across all input datasets.

## Step 5: Normalisation and Batch Correction of Datasets
A common but critical step in phosphoproteomic data analysis is to correct for batch effect. Without batch effect correction, it is often not possible to analyze datasets in an integrative manner. To perform data integration and batch effect correction, we utilise a set of stably phosphorylated sites (SPSs) across a panel of phosphoproteomic datasets (defined from Step 4) and implement normalisation using RUV-III (Molania et al., 2019). To demonstrate batch effect correction, we will perform RUVphospho to normalise a SILAC data from L6 myotubes treated with two factors: 1) AICAR, an analog of adenosine monophosphate (AMP) that stimulates AMPK activity and/or 2) insulin.

CRITICAL: RUV-III requires a complete data matrix. If you have not followed through the steps above, you will need to perform imputation of the missing values. The imputed values are removed by default after normalisation but can be retained for downstream analysis if the users wish to use the imputed matrix.

Load quantification matrix of phosphoproteomic data and prepare PPE object as before
```{r}
load("../Data/PXD019127_ratio_myoblast.RData", verbose = TRUE)
ppe <- PhosphoExperiment(assays = list(Quantification = as.matrix(PXD019127_ratio_myoblast)))

rowNames = strsplit(rownames(ppe), "~")

ppe@GeneSymbol <- toupper(sapply(rowNames, "[[", 2))
ppe@Residue <- gsub("[0-9]","", sapply(rowNames, "[[", 3))
ppe@Site <- as.numeric(gsub("[A-Z]","", sapply(rowNames, "[[", 3)))
ppe@Sequence <- sapply(rowNames, "[[", 4)

# Generate colData
sample_name = strsplit(colnames(ppe), "_")
df = S4Vectors::DataFrame(
    condition = sapply(sample_name, "[[", 1),
    replicate = gsub("exp", "", sapply(sample_name, "[[", 2))
)
rownames(df) = colnames(ppe)
colData(ppe) = df
```

Diagnosing batch effect
```{r}
plotQC(assay(ppe,"Quantification"), panel = "dendrogram", grps = colData(ppe)$condition, labels = colnames(ppe)) + ggplot2::ggtitle("before batch correction")
plotQC(assay(ppe,"Quantification"), panel = "pca", grps = colData(ppe)$condition, labels = colnames(ppe)) + ggplot2::ggtitle("before batch correction")
```

Correcting batch effect
```{r}
design = model.matrix(~ colData(ppe)$condition - 1)
head(design) # observe first 6 rows of the design matrix
sites = paste(
    sapply(ppe@GeneSymbol, function(x)x), ";", 
    sapply(ppe@Residue, function(x)x), 
    sapply(ppe@Site, function(x)x), ";", 
    sep = ""
)
data(SPSs)
ctl = which(sites %in% SPSs)

# Run RUV
ppe = RUVphospho(ppe, M = design, k = 3, ctl = ctl)
```

Visualise QC plots
```{r}
# plot after batch correction
p1 = plotQC(assay(ppe,"Quantification"), grps=colData(ppe)$condition, labels = colnames(ppe), panel = "dendrogram")
p2 = plotQC(assay(ppe,"normalised"), grps=colData(ppe)$condition, labels = colnames(ppe), panel="dendrogram")
ggpubr::ggarrange(p1, p2, nrow = 1)

p1 = plotQC(assay(ppe,"Quantification"), panel = "pca", grps=colData(ppe)$condition, labels = colnames(ppe)) +
  ggplot2::ggtitle("Before Batch correction")
p2 = plotQC(assay(ppe,"normalised"), grps=colData(ppe)$condition, labels = colnames(ppe), panel="pca") +
  ggplot2::ggtitle("After Batch correction")
ggpubr::ggarrange(p1, p2, nrow = 2)
```

We can now save the final processed data for future use
```{r}
# save data
PXD019127_ppe_myoblast = ppe
#save(PXD019127_ppe_myoblast, file = "../Data/PXD019127_ppe_myoblast.RData")
```

Pause Point: This is an ideal pause point as we have generated a fully processed data. By now, you should have a good idea of the data quality and have performed the necessary processing to filter any suboptimal sites, imputed missing values (if present) and diagnosed and corrected any batch effect present in the data. 

## Step 6: Predicting Kinase Substrates
A key end-goal of phosphoproteomic data analysis is to identify kinases that are responsible for the phosphorylation of specific sites. The basic computational approach to annotate kinases to their substrates or phosphosites is to find consensus amino acid sequences around the phosphorylation site. We can go beyond this approach by considering cell type and/or treatment (perturbation) specificity of phosphorylation. PhosR implements a multi-step framework that contains two major components including (i) a kinaseSubstrateScore function which scores a given phosphosite using kinase recognition motif and phosphoproteomic dynamics, and (ii) a kinaseSubstratePred function which synthesise the scores generated from (i) for predicting kinase-substrate relationships using an adaptive sampling-based positive-unlabeled learning method (Yang et al., 2018).

In the original PhosR publication, we demonstrate the application of the scoring method to the myotube phosphoproteome and uncover potential kinase-substrate pairs and global relationships between kinases. We confirm well established substrates of AMPK in our publication: ACACA S79, AKAP1 S103, SMCR8 S488 (Hoffman et al., 2015) and MTFR1L S100 (Schaffer et al., 2015). Importantly, PhosR generates a list of potential candidates not included in the PhosphoSitePlus database for validation.

```{r}
load("../Data/PXD019127_ppe_myoblast.RData", verbose = TRUE)
ppe = PXD019127_ppe_myoblast
mat = SummarizedExperiment::assay(ppe, "normalised")
```

CRITICAL: If you have not processed your data, please go through Steps 1 to 5 to perform the necessary processing prior to performing the downstream analysis in Steps 6 and 7.

Filter for up-regulated phosphosites
```{r}
mat.mean <- meanAbundance(mat, grps = colData(ppe)$condition)
aov <- matANOVA(mat=mat, grps=colData(ppe)$condition)
idx <- (aov < 0.05) & (rowSums(mat.mean > 0.5) > 0)
mat.reg <- mat[idx, ,drop = FALSE]
```

Standardise the matrix
```{r}
mat.std <- PhosR::standardise(mat.reg)
rownames(mat.std) <- sapply(strsplit(rownames(mat.std), "~"), function(x){gsub(" ", "", paste(toupper(x[2]), x[3], "", sep=";"))})
```

CRITICAL: To calculate the profile matching score, we rely on the z-score transformed matrix to compare the profiles of phosphosites. Thus, the standardization step is critical.

Kinase substrate scoring step integrates information from both kinase recognition motif (i.e., motif matching score) and experimental perturbation (i.e., profile matching score) for prioritising kinases that may be regulating the phosphorylation level of each site quantified in the 

Run PhosR kinase-substrate prediction with the default parameters. PhosR generates the final combined scores of the motif matching score and the profile matching score by taking into account the number of sequences and substrates used for calculating the motif and profile of the kinase. 
```{r}
data("KinaseMotifs")
seqs <- ppe@Sequence[idx]
kssMat <- kinaseSubstrateScore(substrate.list = PhosphoSite.mouse, 
                               mat = mat.std, seqs = seqs,
                               numMotif = 5, numSub = 1, verbose = FALSE)
```

As the second and last step of kinase-substrate prediction, PhosR uses the ‘kinaseSubstratePred’ function to synthesise the scores generated from ‘kinaseSubstrateScore’ to predict the kinase-substrate relationships using an adaptive sampling-based positive-unlabeled learning method (Yang et al., 2018). This step prioritises the kinases that are most likely to regulate a phosphosite.

```{r}
set.seed(1)
predMat <- kinaseSubstratePred(kssMat, top = 30, verbose = FALSE)
colnames(predMat)
```

Pause Point: Once the desired parameters are used to successfully run the kinase-substrate prediction, you may find this to be a good place to pause and save the results before proceeding with the visualisation steps.

## Step 7: Constructing Signalling Networks (Signalomes)

Constructing signaling networks referred to as “Signalomes” is a useful feature in PhosR that allows the users to obtain a global view of kinase regulation and to establish distinct modules of proteins that demonstrate similar kinase and dynamic regulation upon perturbation or across a time-course.

An important feature of PhosR signalomes is that the resultant signaling modules denote a set of proteins and not phosphosites. Proteins are frequently phosphorylated at multiple sites and these sites may not necessarily be targeted by the same kinases. Site- and protein-centric analyses of phosphoproteomic data lie at opposite ends of the spectrum, with the former treating phosphosites on the same protein independently and ignoring the host protein information, and the latter focussing on a specific protein, losing information from individual phosphosites. Using our global kinase-substrate scoring of phosphosites, we generate signalomes wherein dynamic changes in phosphorylation within and across proteins are conjointly analysed, allowing us to detect proteins that are co-regulated across multiple phosphosites.

The signalome construction uses the outputs of ‘kinaseSubstrateScore’and ‘kinaseSubstratePred’ functions for the generation of a visualisation of the kinase regulation of discrete regulatory protein modules present in our phosphoproteomic data.

To generate the signalomes, run:
```{r message = FALSE}
kinaseOI = c("AKT1")
signalomesRes <- Signalomes(KSR = kssMat, 
                            predMatrix = predMat, 
                            exprsMat = mat.std, 
                            module_res = 6,
                            KOI = kinaseOI)
```

We can visualise the signalomes as a balloon plot. Using the resulting visualisation, we are better able to compare the differences in kinase regulation of the modules and the varying proportions of regulation. In the balloon plot, the size of the balloons denotes the percentage magnitude of kinase regulation in each module.
```{r message = FALSE}
### generate palette
my_color_palette <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(8, "Accent"))
kinase_all_color <- my_color_palette(ncol(kssMat$combinedScoreMatrix))
names(kinase_all_color) <- colnames(kssMat$combinedScoreMatrix)
kinase_signalome_color <- kinase_all_color[colnames(predMat)]

plotSignalomeMap(signalomes = signalomesRes, color = kinase_signalome_color)
```

Finally, we can also plot the signalome network that illustrates the connectivity between kinase signalome networks.
```{r message = FALSE}
plotKinaseNetwork(KSR = kssMat, predMatrix = predMat, threshold = 0.95, color = kinase_all_color)
```

# Session Info
```{r}
utils::sessionInfo()
```

